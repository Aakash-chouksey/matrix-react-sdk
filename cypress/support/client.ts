/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import type { ICreateRoomOpts } from "matrix-js-sdk/src/@types/requests";
import type { MatrixClient } from "matrix-js-sdk/src/client";
import type { Room } from "matrix-js-sdk/src/models/room";
import type { IJoinRoomOpts } from "matrix-js-sdk/src/@types/requests";
import Chainable = Cypress.Chainable;

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the MatrixClient from the MatrixClientPeg
             */
            getClient(): Chainable<MatrixClient | undefined>;
            /**
             * Create a room with given options.
             * @param options the options to apply when creating the room
             * @return the ID of the newly created room
             */
            createRoom(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Invites the given user to the given room.
             * @param roomId the id of the room to invite to
             * @param userId the id of the user to invite
             */
            inviteUser(roomId: string, userId: string): Chainable<{}>;
            /**
             * Joins the current user to the given room.
             * @param roomId the id of the room to join
             * @param opts the options when joining a room
             */
            joinRoom(roomId: string, opts?: IJoinRoomOpts): Chainable<{}>;
            /**
             * Waits for the given room to be synced locally
             * @param roomId the id of the room to wait for locally
             */
            waitForRoom(roomId: string): Chainable<string>;
        }
    }
}

Cypress.Commands.add("getClient", (): Chainable<MatrixClient | undefined> => {
    return cy.window({ log: false }).then(win => win.mxMatrixClientPeg.matrixClient);
});

Cypress.Commands.add("createRoom", (options: ICreateRoomOpts): Chainable<string> => {
    cy.window({ log: false })
        .then(async win => {
            const cli = win.mxMatrixClientPeg.matrixClient;
            const resp = await cli.createRoom(options);
            const roomId = resp.room_id;
            return roomId;
        })
        .as('roomId');

    return cy.get<string>("@roomId").then(roomId => {
        return cy.waitForRoom(roomId);
    });
});

Cypress.Commands.add("inviteUser", (roomId: string, userId: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.invite(roomId, userId);
    });
});

Cypress.Commands.add("joinRoom", (roomId: string, opts?: IJoinRoomOpts): Chainable<{}> => {
    cy.getClient().then(async (cli: MatrixClient) => {
        return cli.joinRoom(roomId, opts);
    });

    // Wait for the room to be available locally
    return cy.waitForRoom(roomId);
});

Cypress.Commands.add("waitForRoom", (roomId: string): Chainable<string> => {
    return cy.window({ log: false }).then(async win => {
        const cli = win.mxMatrixClientPeg.matrixClient;

        // Wait for the room to be available locally
        if (!cli.getRoom(roomId)) {
            await new Promise<void>(resolve => {
                const onRoom = (room: Room) => {
                    if (room.roomId === roomId) {
                        cli.off(win.matrixcs.ClientEvent.Room, onRoom);
                        resolve();
                    }
                };
                cli.on(win.matrixcs.ClientEvent.Room, onRoom);
            });
        }

        return roomId;
    });
});
