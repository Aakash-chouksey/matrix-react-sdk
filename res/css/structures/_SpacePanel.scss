/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

$topLevelHeight: 32px;
$nestedHeight: 24px;
$gutterSize: 21px;
$activeStripeSize: 4px;

.mx_SpacePanel {
    flex: 0 0 auto;
    background-color: $groupFilterPanel-bg-color;

    // Create another flexbox so the Panel fills the container
    display: flex;
    flex-direction: column;
    overflow-y: auto;

    ul {
        margin: 0;
        list-style: none;
        padding: 0;
        padding-left: 16px;
    }

    .mx_AutoHideScrollbar {
        padding: 16px 12px 16px 0;
    }

    .mx_SpaceButton_toggleCollapse {
        cursor: pointer;
    }

    .mx_SpaceItem {
        position: relative;
    }

    .mx_SpaceItem.collapsed {
        & > .mx_SpaceButton > .mx_SpaceButton_toggleCollapse {
            transform: rotate(-90deg);
        }

        & > .mx_SpaceTreeLevel {
            display: none;
        }
    }

    .mx_SpaceItem:not(.hasSubSpaces) > .mx_SpaceButton  {
        margin-left: $gutterSize;

        &.mx_SpaceButton_active {
            &::before {
                left: -$gutterSize;
            }
        }
    }

    .mx_SpacePanel_newWrapper {
        margin-left: $gutterSize;
    }

    .mx_SpaceButton {
        border-radius: 8px;
        position: relative;
        margin-bottom: 16px;
        display: flex;
        align-items: center;

        .mx_SpaceButton_name {
            flex: 1;
            margin-left: 8px;
            white-space: nowrap;
            display: block;
            max-width: 150px;
            text-overflow: ellipsis;
            overflow: hidden;
        }

        .mx_SpaceButton_toggleCollapse {
            width: calc($gutterSize - $activeStripeSize);
            margin-left: $activeStripeSize;
            height: $gutterSize;
            mask-position: center;
            mask-size: 21px;
            mask-repeat: no-repeat;
            background-color: $roomlist-header-color;
            mask-image: url('$(res)/img/feather-customised/chevron-down.svg');
        }

        &.mx_SpaceButton_active {
            &::before {
                position: absolute;
                content: '';
                width: $activeStripeSize;
                top: 0;
                left: 0;
                bottom: 0;
                background-color: $accent-color;
                border-radius: 0 4px 4px 0;
            }
        }

        &.mx_SpaceButton_home, &.mx_SpaceButton_new {
            width: $topLevelHeight;
            height: $topLevelHeight;

            &::after {
                position: absolute;
                content: '';
                width: $topLevelHeight;
                height: $topLevelHeight;
                top: 0;
                left: 0;
                mask-position: center;
                mask-repeat: no-repeat;
            }
        }

        &.mx_SpaceButton_home {
            background-color: #ffffff;

            &::after {
                background-color: #3F3D3D;
                mask-image: url('$(res)/img/element-icons/home.svg');
                mask-size: 18px;
            }
        }

        &.mx_SpaceButton_new {
            background-color: $accent-color;
            transition: all .1s ease-in-out; // TODO transition

            &::after {
                background-color: #ffffff;
                mask-image: url('$(res)/img/element-icons/plus.svg');
                mask-size: 14px;
                transition: all .2s ease-in-out; // TODO transition
            }
        }

        &.mx_SpaceButton_newCancel {
            background-color: $icon-button-color;

            &::after {
                transform: rotate(45deg);
            }
        }

        .mx_BaseAvatar_image {
            border-radius: 8px;
        }

        .mx_NotificationBadge {
            position: absolute;
            right: -4px;
            top: -4px;
        }
    }


    /* root space buttons are bigger and not indented */
    & > .mx_AutoHideScrollbar {
        & > .mx_SpaceButton {
            height: $topLevelHeight;

            &.mx_SpaceButton_active::before {
                height: $topLevelHeight;
            }
        }

        & > ul {
            padding-left: 0;
        }
    }

}
