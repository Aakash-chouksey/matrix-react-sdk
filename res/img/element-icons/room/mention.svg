<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M15.4846 16.3684C15.4542 16.3888 15.4242 16.4103 15.3948 16.4329C13.3349 18.0103 10.4597 17.9609 8.45528 16.3135C6.45085 14.6662 5.84542 11.8552 6.99397 9.52884C8.14254 7.20246 10.7424 5.97384 13.2691 6.56348C15.7957 7.1531 17.5831 9.40556 17.5833 12C17.5833 12 17.5833 12 17.5833 12V12.7332C17.5833 12.8998 17.3742 13.1831 17.0417 13.1831C16.7094 13.1831 16.5 12.8997 16.5 12.7329L16.5 11.9999L16.5 11.9988L16.5 11.996L16.5 11.9956L15.75 11.9967C15.7482 9.92713 14.07 8.25 12 8.25C9.92893 8.25 8.25 9.92893 8.25 12C8.25 14.0711 9.92893 15.75 12 15.75C12.9106 15.75 13.7454 15.4254 14.395 14.8857L15.4846 16.3684ZM15.4846 16.3684C15.1184 16.2155 14.7759 16.0107 14.4693 15.7624L15.4846 16.3684ZM21.0833 11.9998V11.9998C21.0831 7.77882 18.175 4.11432 14.0645 3.15506C9.95394 2.1958 5.72424 4.19458 3.85563 7.9794C1.98701 11.7642 2.97202 16.3375 6.23302 19.0175C9.49402 21.6976 14.1715 21.778 17.5227 19.2117C18.2901 18.6241 18.4358 17.5256 17.8482 16.7583C17.8115 16.7104 17.7729 16.665 17.7324 16.622C19.5385 16.3018 21.0833 14.7552 21.0833 12.7332V11.9998ZM12 11C12.5523 11 13 11.4477 13 12C13 12.5523 12.5523 13 12 13C11.4477 13 11 12.5523 11 12C11 11.4477 11.4477 11 12 11ZM1.75 12C1.75 6.33908 6.33908 1.75 12 1.75C17.6609 1.75 22.25 6.33908 22.25 12C22.25 17.6609 17.6609 22.25 12 22.25C6.33908 22.25 1.75 17.6609 1.75 12Z" fill="currentColor" stroke="currentColor" stroke-width="1.5" stroke-linecap="round"/>
</svg>
