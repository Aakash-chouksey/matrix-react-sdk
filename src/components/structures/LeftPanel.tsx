/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import * as React from "react";
import { createRef } from "react";
import classNames from "classnames";
import { Room } from "matrix-js-sdk/src/models/room";

import GroupFilterPanel from "./GroupFilterPanel";
import CustomRoomTagPanel from "./CustomRoomTagPanel";
import dis from "../../dispatcher/dispatcher";
import { _t } from "../../languageHandler";
import RoomList from "../views/rooms/RoomList";
import { Action } from "../../dispatcher/actions";
import UserMenu from "./UserMenu";
import RoomSearch from "./RoomSearch";
import RoomBreadcrumbs from "../views/rooms/RoomBreadcrumbs";
import { BreadcrumbsStore } from "../../stores/BreadcrumbsStore";
import { UPDATE_EVENT } from "../../stores/AsyncStore";
import ResizeNotifier from "../../utils/ResizeNotifier";
import SettingsStore from "../../settings/SettingsStore";
import RoomListStore, { LISTS_UPDATE_EVENT } from "../../stores/room-list/RoomListStore";
import IndicatorScrollbar from "../structures/IndicatorScrollbar";
import AccessibleTooltipButton from "../views/elements/AccessibleTooltipButton";
import { OwnProfileStore } from "../../stores/OwnProfileStore";
import RoomListNumResults from "../views/rooms/RoomListNumResults";
import LeftPanelWidget from "./LeftPanelWidget";
import {replaceableComponent} from "../../utils/replaceableComponent";
import {mediaFromMxc} from "../../customisations/Media";
import SpaceStore, {UPDATE_SELECTED_SPACE} from "../../stores/SpaceStore";
import { getKeyBindingsManager, RoomListAction } from "../../KeyBindingsManager";

interface IProps {
    isMinimized: boolean;
    resizeNotifier: ResizeNotifier;
}

interface IState {
    showBreadcrumbs: boolean;
    showGroupFilterPanel: boolean;
    activeSpace?: Room;
}

// List of CSS classes which should be included in keyboard navigation within the room list
const cssClasses = [
    "mx_RoomSearch_input",
    "mx_RoomSearch_minimizedHandle", // minimized <RoomSearch />
    "mx_RoomSublist_headerText",
    "mx_RoomTile",
    "mx_RoomSublist_showNButton",
];

@replaceableComponent("structures.LeftPanel")
export default class LeftPanel extends React.Component<IProps, IState> {
    private listContainerRef: React.RefObject<HTMLDivElement> = createRef();
    private groupFilterPanelWatcherRef: string;
    private bgImageWatcherRef: string;
    private focusedElement = null;
    private isDoingStickyHeaders = false;

    constructor(props: IProps) {
        super(props);

        this.state = {
            showBreadcrumbs: BreadcrumbsStore.instance.visible,
            showGroupFilterPanel: SettingsStore.getValue('TagPanel.enableTagPanel'),
            activeSpace: SpaceStore.instance.activeSpace,
        };

        BreadcrumbsStore.instance.on(UPDATE_EVENT, this.onBreadcrumbsUpdate);
        RoomListStore.instance.on(LISTS_UPDATE_EVENT, this.onBreadcrumbsUpdate);
        OwnProfileStore.instance.on(UPDATE_EVENT, this.onBackgroundImageUpdate);
        SpaceStore.instance.on(UPDATE_SELECTED_SPACE, this.updateActiveSpace);
        this.bgImageWatcherRef = SettingsStore.watchSetting(
            "RoomList.backgroundImage", null, this.onBackgroundImageUpdate);
        this.groupFilterPanelWatcherRef = SettingsStore.watchSetting("TagPanel.enableTagPanel", null, () => {
            this.setState({showGroupFilterPanel: SettingsStore.getValue("TagPanel.enableTagPanel")});
        });
    }

    public componentWillUnmount() {
        SettingsStore.unwatchSetting(this.groupFilterPanelWatcherRef);
        SettingsStore.unwatchSetting(this.bgImageWatcherRef);
        BreadcrumbsStore.instance.off(UPDATE_EVENT, this.onBreadcrumbsUpdate);
        RoomListStore.instance.off(LISTS_UPDATE_EVENT, this.onBreadcrumbsUpdate);
        OwnProfileStore.instance.off(UPDATE_EVENT, this.onBackgroundImageUpdate);
        SpaceStore.instance.off(UPDATE_SELECTED_SPACE, this.updateActiveSpace);
    }

    private updateActiveSpace = (activeSpace: Room) => {
        this.setState({ activeSpace });
    };

    private onExplore = () => {
        dis.fire(Action.ViewRoomDirectory);
    };

    private onBreadcrumbsUpdate = () => {
        const newVal = BreadcrumbsStore.instance.visible;
        if (newVal !== this.state.showBreadcrumbs) {
            this.setState({showBreadcrumbs: newVal});
        }
    };

    private onBackgroundImageUpdate = () => {
        // Note: we do this in the LeftPanel as it uses this variable most prominently.
        const avatarSize = 32; // arbitrary
        let avatarUrl = OwnProfileStore.instance.getHttpAvatarUrl(avatarSize);
        const settingBgMxc = SettingsStore.getValue("RoomList.backgroundImage");
        if (settingBgMxc) {
            avatarUrl = mediaFromMxc(settingBgMxc).getSquareThumbnailHttp(avatarSize);
        }

        const avatarUrlProp = `url(${avatarUrl})`;
        if (!avatarUrl) {
            document.body.style.removeProperty("--avatar-url");
        } else if (document.body.style.getPropertyValue("--avatar-url") !== avatarUrlProp) {
            document.body.style.setProperty("--avatar-url", avatarUrlProp);
        }
    };

    private onFocus = (ev: React.FocusEvent) => {
        this.focusedElement = ev.target;
    };

    private onBlur = () => {
        this.focusedElement = null;
    };

    private onKeyDown = (ev: React.KeyboardEvent) => {
        if (!this.focusedElement) return;

        const action = getKeyBindingsManager().getRoomListAction(ev);
        switch (action) {
            case RoomListAction.NextRoom:
            case RoomListAction.PrevRoom:
                ev.stopPropagation();
                ev.preventDefault();
                this.onMoveFocus(action === RoomListAction.PrevRoom);
                break;
        }
    };

    private selectRoom = () => {
        const firstRoom = this.listContainerRef.current.querySelector<HTMLDivElement>(".mx_RoomTile");
        if (firstRoom) {
            firstRoom.click();
            return true; // to get the field to clear
        }
    };

    private onMoveFocus = (up: boolean) => {
        let element = this.focusedElement;

        let descending = false; // are we currently descending or ascending through the DOM tree?
        let classes: DOMTokenList;

        do {
            const child = up ? element.lastElementChild : element.firstElementChild;
            const sibling = up ? element.previousElementSibling : element.nextElementSibling;

            if (descending) {
                if (child) {
                    element = child;
                } else if (sibling) {
                    element = sibling;
                } else {
                    descending = false;
                    element = element.parentElement;
                }
            } else {
                if (sibling) {
                    element = sibling;
                    descending = true;
                } else {
                    element = element.parentElement;
                }
            }

            if (element) {
                classes = element.classList;
            }
        } while (element && (!cssClasses.some(c => classes.contains(c)) || element.offsetParent === null));

        if (element) {
            element.focus();
            this.focusedElement = element;
        }
    };

    private renderHeader(): React.ReactNode {
        return (
            <div className="mx_LeftPanel_userHeader">
                <UserMenu isMinimized={this.props.isMinimized} />
            </div>
        );
    }

    private renderBreadcrumbs(): React.ReactNode {
        if (this.state.showBreadcrumbs && !this.props.isMinimized) {
            return (
                <IndicatorScrollbar
                    className="mx_LeftPanel_breadcrumbsContainer mx_AutoHideScrollbar"
                    verticalScrollsHorizontally={true}
                    // Firefox sometimes makes this element focusable due to
                    // overflow:scroll;, so force it out of tab order.
                    tabIndex={-1}
                >
                    <RoomBreadcrumbs />
                </IndicatorScrollbar>
            );
        }
    }

    private renderSearchExplore(): React.ReactNode {
        return (
            <div
                className="mx_LeftPanel_filterContainer"
                onFocus={this.onFocus}
                onBlur={this.onBlur}
                onKeyDown={this.onKeyDown}
            >
                <RoomSearch
                    isMinimized={this.props.isMinimized}
                    onKeyDown={this.onKeyDown}
                    onSelectRoom={this.selectRoom}
                />
                <AccessibleTooltipButton
                    className={classNames("mx_LeftPanel_exploreButton", {
                        mx_LeftPanel_exploreButton_space: !!this.state.activeSpace,
                    })}
                    onClick={this.onExplore}
                    title={_t("Explore rooms")}
                />
            </div>
        );
    }

    public render(): React.ReactNode {
        let leftLeftPanel;
        if (this.state.showGroupFilterPanel) {
            leftLeftPanel = (
                <div className="mx_LeftPanel_GroupFilterPanelContainer">
                    <GroupFilterPanel />
                    {SettingsStore.getValue("feature_custom_tags") ? <CustomRoomTagPanel /> : null}
                </div>
            );
        }

        const roomList = <RoomList
            onKeyDown={this.onKeyDown}
            resizeNotifier={this.props.resizeNotifier}
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            isMinimized={this.props.isMinimized}
            activeSpace={this.state.activeSpace}
        />;

        const containerClasses = classNames({
            "mx_LeftPanel": true,
            "mx_LeftPanel_minimized": this.props.isMinimized,
        });

        const roomListClasses = classNames(
            "mx_LeftPanel_actualRoomListContainer",
            "mx_AutoHideScrollbar",
        );

        return (
            <div className={containerClasses}>
                {leftLeftPanel}
                <aside className="mx_LeftPanel_roomListContainer">
                    {this.renderHeader()}
                    {this.renderSearchExplore()}
                    {this.renderBreadcrumbs()}
                    <RoomListNumResults />
                    <div className="mx_LeftPanel_roomListWrapper">
                        <div
                            className={roomListClasses}
                            ref={this.listContainerRef}
                            // Firefox sometimes makes this element focusable due to
                            // overflow:scroll;, so force it out of tab order.
                            tabIndex={-1}
                        >
                            {roomList}
                        </div>
                    </div>
                    { !this.props.isMinimized && <LeftPanelWidget /> }
                </aside>
            </div>
        );
    }
}
