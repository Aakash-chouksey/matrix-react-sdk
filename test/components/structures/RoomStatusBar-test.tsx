/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { act } from 'react-dom/test-utils';
import { Room } from "matrix-js-sdk/src/models/room";
import { PendingEventOrdering } from 'matrix-js-sdk/src/client';
import { EventStatus, MatrixEvent } from "matrix-js-sdk/src/models/event";
import {
    ISyncStateData, SyncState,
} from 'matrix-js-sdk/src/sync';
import { MatrixError } from "matrix-js-sdk/src/http-api";

import MatrixClientContext from "../../../src/contexts/MatrixClientContext";
import RoomStatusBar from "../../../src/components/structures/RoomStatusBar";
import { getUnsentMessages } from "../../../src/components/structures/RoomStatusBar";
import { MatrixClientPeg } from "../../../src/MatrixClientPeg";
import { mkEvent, stubClient } from "../../test-utils/test-utils";
import { mkThread } from "../../test-utils/threads";
import { sleep } from "matrix-js-sdk/src/utils";

// Fake date to give a predictable snapshot
const realDateNow = Date.now;
const realDateToISOString = Date.prototype.toISOString;
Date.now = jest.fn(() => 2345678901234);
// eslint-disable-next-line no-extend-native
Date.prototype.toISOString = jest.fn(() => "2021-11-23T14:35:14.240Z");

afterAll(() => {
    Date.now = realDateNow;
    // eslint-disable-next-line no-extend-native
    Date.prototype.toISOString = realDateToISOString;
});

describe("RoomStatusBar", () => {
    let client;
    beforeEach(() => {
        stubClient();
        client = MatrixClientPeg.get();
    });

    const customRender = (room: Room) => {
        return render(
            <MatrixClientContext.Provider value={client}>
                <RoomStatusBar room={room} />
            </MatrixClientContext.Provider>,
        );
    };

    describe("getUnsentMessages", () => {
        const ROOM_ID = "!roomId:example.org";
        let room: Room;
        let event: MatrixEvent;

        beforeEach(() => {
            jest.clearAllMocks();

            room = new Room(ROOM_ID, client, client.getUserId(), {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });
            event = mkEvent({
                event: true,
                type: "m.room.message",
                user: "@user1:server",
                room: "!room1:server",
                content: {},
            });
            event.status = EventStatus.NOT_SENT;
        });

        it("returns no unsent messages", () => {
            expect(getUnsentMessages(room)).toHaveLength(0);
        });

        it("checks the event status", () => {
            room.addPendingEvent(event, "123");

            expect(getUnsentMessages(room)).toHaveLength(1);
            event.status = EventStatus.SENT;

            expect(getUnsentMessages(room)).toHaveLength(0);
        });

        it("only returns events related to a thread", () => {
            room.addPendingEvent(event, "123");

            const { rootEvent, events } = mkThread({
                room,
                client,
                authorId: "@alice:example.org",
                participantUserIds: ["@alice:example.org"],
                length: 2,
            });
            rootEvent.status = EventStatus.NOT_SENT;
            room.addPendingEvent(rootEvent, rootEvent.getId());
            for (const event of events) {
                event.status = EventStatus.NOT_SENT;
                room.addPendingEvent(event, Date.now() + Math.random() + "");
            }

            const pendingEvents = getUnsentMessages(room, rootEvent.getId());

            expect(pendingEvents[0].threadRootId).toBe(rootEvent.getId());
            expect(pendingEvents[1].threadRootId).toBe(rootEvent.getId());
            expect(pendingEvents[2].threadRootId).toBe(rootEvent.getId());

            // Filters out the non thread events
            expect(pendingEvents.every(ev => ev.getId() !== event.getId())).toBe(true);
        });
    });

    it('does not show anything when no sync error or other status', () => {
        const r1 = new Room("r1", client, "@name:example.com", {
            pendingEventOrdering: PendingEventOrdering.Detached,
        });

        const wrapper = customRender(r1);
        expect(wrapper.asFragment()).toMatchSnapshot();
    });

    describe('connectivity lost bar', () => {
        it('should show connection lost bar when sync has an error', () => {
            client.getSyncState = (): SyncState => SyncState.Error,
            client.getSyncStateData = (): ISyncStateData => ({
                error: new MatrixError({
                    errcode: 'FAKE_ERROR',
                    error: "Fake sync error",
                }),
            });

            const r1 = new Room("r1", client, "@name:example.com", {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });

            const wrapper = customRender(r1);
            expect(wrapper.asFragment()).toMatchSnapshot();
        });

        it('connectivity lost bar has priority over the timeline refresh bar', () => {
            // Show connectivity lost bar
            client.getSyncState = (): SyncState => SyncState.Error,
            client.getSyncStateData = (): ISyncStateData => ({
                error: new MatrixError({
                    errcode: 'FAKE_ERROR',
                    error: "Fake sync error",
                }),
            });

            const r1 = new Room("r1", client, "@name:example.com", {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });

            // Show timeline needs refresh bar
            r1.setTimelineNeedsRefresh(true);

            const wrapper = customRender(r1);
            expect(wrapper.asFragment()).toMatchSnapshot();
        });
    });

    describe('timeline needs refresh bar (history import)', () => {
        it('should show timeline refresh bar when history import detected', () => {
            const r1 = new Room("r1", client, "@name:example.com", {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });
            // Show timeline needs refresh bar
            r1.setTimelineNeedsRefresh(true);

            const wrapper = customRender(r1);
            expect(wrapper.asFragment()).toMatchSnapshot();
        });

        it('should refresh timeline for room when button clicked', () => {
            const r1 = new Room("r1", client, "@name:example.com", {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });
            // Show timeline needs refresh bar
            r1.setTimelineNeedsRefresh(true);

            r1.refreshLiveTimeline = jest.fn();

            const wrapper = customRender(r1);

            act(() => {
                fireEvent.click(wrapper.getByTestId('refresh-timeline-button'));
            });

            // Make sure that the SDK was called to refresh the timeline
            expect(r1.refreshLiveTimeline).toHaveBeenCalled();
        });

        it('should show error state with option to submit debug logs ' +
           'in timeline refresh bar when something went wrong while refreshing', async () => {
            const r1 = new Room("r1", client, "@name:example.com", {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });
            // Show timeline needs refresh bar
            r1.setTimelineNeedsRefresh(true);

            const wrapper = customRender(r1);

            r1.refreshLiveTimeline = jest.fn().mockRejectedValue(new Error('Fake error in test'));
            act(() => {
                fireEvent.click(wrapper.getByTestId('refresh-timeline-button'));
            });

            // Make sure that the SDK was called to refresh the timeline
            expect(r1.refreshLiveTimeline).toHaveBeenCalled();

            // Expect error to be shown. We have to wait for the UI to transition.
            await waitFor(() => expect(wrapper.getByTestId('historical-import-detected-error-content')).toBeDefined());

            // Expect an option to submit debug logs to be shown
            expect(wrapper.getByTestId('historical-import-detected-error-submit-debug-logs-button')).toBeDefined();
        });

        it('should show error state without submit debug logs option ' +
           'in timeline refresh bar when ConnectionError while refreshing', async () => {
            const r1 = new Room("r1", client, "@name:example.com", {
                pendingEventOrdering: PendingEventOrdering.Detached,
            });
            // Show timeline needs refresh bar
            r1.setTimelineNeedsRefresh(true);

            const wrapper = customRender(r1);

            const connectionError = new Error('Fake connection error in test');
            connectionError.name = "ConnectionError";
            r1.refreshLiveTimeline = jest.fn().mockRejectedValue(connectionError);
            act(() => {
                fireEvent.click(wrapper.getByTestId('refresh-timeline-button'));
            });

            // Make sure that the SDK was called to refresh the timeline
            expect(r1.refreshLiveTimeline).toHaveBeenCalled();

            // Expect error to be shown
            await waitFor(() => expect(wrapper.getByTestId('historical-import-detected-error-content')).toBeDefined());

            // The submit debug logs option should *NOT* be shown.
            //
            // We have to use `queryBy` so that it can return `null` for something that does not exist.
            expect(wrapper.queryByTestId('historical-import-detected-error-submit-debug-logs-button]')).toBeNull();
        });
    });
});
