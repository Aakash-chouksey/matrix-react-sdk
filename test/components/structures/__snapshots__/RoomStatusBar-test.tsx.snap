// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomStatusBar connectivity lost bar connectivity lost bar has priority over the timeline refresh bar 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomStatusBar"
  >
    <div
      role="alert"
    >
      <div
        class="mx_RoomStatusBar_connectionLostBar"
      >
        <img
          alt=""
          height="24"
          src="image-file-stub"
          width="24"
        />
        <div>
          <div
            class="mx_RoomStatusBar_connectionLostBar_title"
          >
            Connectivity to the server has been lost.
          </div>
          <div
            class="mx_RoomStatusBar_connectionLostBar_desc"
          >
            Sent messages will be stored until your connection has returned.
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RoomStatusBar connectivity lost bar should show connection lost bar when sync has an error 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomStatusBar"
  >
    <div
      role="alert"
    >
      <div
        class="mx_RoomStatusBar_connectionLostBar"
      >
        <img
          alt=""
          height="24"
          src="image-file-stub"
          width="24"
        />
        <div>
          <div
            class="mx_RoomStatusBar_connectionLostBar_title"
          >
            Connectivity to the server has been lost.
          </div>
          <div
            class="mx_RoomStatusBar_connectionLostBar_desc"
          >
            Sent messages will be stored until your connection has returned.
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RoomStatusBar does not show anything when no sync error or other status 1`] = `<DocumentFragment />`;

exports[`RoomStatusBar timeline needs refresh bar (history import) should show timeline refresh bar when history import detected 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomStatusBar mx_RoomStatusBar_unsentMessages"
    data-testid="historical-import-detected-status-bar"
  >
    <div
      role="alert"
    >
      <div
        class="mx_RoomStatusBar_unsentBadge"
      >
        <img
          alt=""
          height="24"
          src="image-file-stub"
          width="24"
        />
      </div>
      <div>
        <div
          class="mx_RoomStatusBar_unsentTitle"
        >
          History import detected.
        </div>
        <div
          class="mx_RoomStatusBar_unsentDescription"
        >
          History was just imported somewhere in the room. In order to see the historical messages, refresh your timeline.
        </div>
      </div>
      <div
        class="mx_RoomStatusBar_unsentButtonBar"
      >
        <div
          class="mx_AccessibleButton mx_RoomStatusBar_refreshTimelineBtn"
          data-testid="refresh-timeline-button"
          role="button"
          tabindex="0"
        >
          Refresh timeline
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
